streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -960,7 +1011,7 @@ private void removeRevokedTasksFromStateUpdater(final Set<TopicPartition> remain
             for (final Task restoringTask : stateUpdater.getTasks()) {
                 if (restoringTask.isActive()) {
                     if (remainingRevokedPartitions.containsAll(restoringTask.inputPartitions())) {
-                        tasks.addPendingTaskToCloseClean(restoringTask.id());
+                        tasks.addPendingActiveTaskToSuspend(restoringTask.id());
```

Comment:
noob question: Why do we suspend instead of close?

Reply:
A revoked active tasks could be immediately reassigned (i.e. in the very next assignment) as an active task or a standby task to the stream thread. If we closed the revoked active task, we would need to re-create the reassigned active or standby task. If we suspend the revoked active task, we can resume it if it is immediately reassigned as an active task and we can recycle it if it is immediately reassigned as a standby task.

Reply:
I think we should remove the task from other pending sets if it was in them. Since in handling removed / restored tasks from state updater checking the other pending sets (https://github.com/apache/kafka/pull/12600/files#diff-8baa5d7209fc00074bf3fe24d709c2dcf2a44c1623d7ced8c0e29c1d832a3bcbR804) has a higher order, right?

I was thinking about the following steps:

T1: a first rebalance completes where a task was recycled to active, it was added to pending-recycle, and call to `stateUpdater.remove`.
T2: a new rebalance starts, where we `handleRevocation`, and then this task is revoked, we add it to the pending-suspend sets as well
T3: finally the task is returned from state updater, we would check the pending-recycle first and trigger that logic, which would be wrong.

More generally, I think as a principle, a task should only be in one of the pending sets in Tasks. So whenever we are adding a a task to a pending set, we should make sure they are not in other pending sets.

Reply:
Yeah, I had similar thoughts about that a task must only be in one of the bookkeeping sets. I even thought about a dedicated data structure for that. Let me create a PR for it.

Reply:
I introduced one map for all pending sets in `Tasks`. Now a task can only be part of the latest pending set it was assigned to.